XSLImportRule and XSLStyleSheet should use weak pointers instead of raw pointers

rniwa opened the pull request:
#### 896fddc4ea9a19ef7d322097501f479ba0604a80
<pre>
XSLImportRule and XSLStyleSheet should use weak pointers instead of raw pointers
<a href="https://bugs.webkit.org/show_bug.cgi?id=244181">https://bugs.webkit.org/show_bug.cgi?id=244181</a>

Reviewed by Darin Adler.

Deploy WeakPtr where raw pointers are used.

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::setXSLStyleSheet):
* Source/WebCore/xml/XSLImportRule.cpp:
(WebCore::XSLImportRule::XSLImportRule):
(WebCore::XSLImportRule::setXSLStyleSheet):
(WebCore::XSLImportRule::loadSheet):
* Source/WebCore/xml/XSLImportRule.h:
(WebCore::XSLImportRule::parentStyleSheet const):
* Source/WebCore/xml/XSLStyleSheet.h:
* Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
(WebCore::XSLStyleSheet::XSLStyleSheet):
(WebCore::XSLStyleSheet::checkLoaded):
(WebCore::XSLStyleSheet::loadChildSheet):
(WebCore::XSLStyleSheet::ownerDocument):
(WebCore::XSLStyleSheet::locateStylesheetSubResource):

Canonical link: <a href="https://commits.webkit.org/253695@main">https://commits.webkit.org/253695@main</a>
</pre>

<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57d92473bb889afaa874fa2b1a1b4a1058cf57e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86769 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95601 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149350 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90750 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29212 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25596 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78936 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90838 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23588 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73665 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23633 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78907 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66640 "Found 1 new API test failure: /WebKit2Gtk/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26975 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12761 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26898 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13775 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36640 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1029 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28524 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33054 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/xml/XSLImportRule.cpp:
```diff
@@ -53,11 +51,9 @@ void XSLImportRule::setXSLStyleSheet(const String& href, const URL& baseURL, con
-    XSLStyleSheet* parent = parentStyleSheet();
-    if (parent)
-        m_styleSheet->setParentStyleSheet(parent);
+    // FIXME: parentStyleSheet() should never be null here.
```
I don’t understand why this comment says `FIXME`. What are we going to fix?

rniwa replied:
Ideally, we pass a reference instead of a pointer if it's never null. I can't easily validate that, however.

darinadler commented on Source/WebCore/xml/XSLImportRule.cpp:
```diff
@@ -73,27 +69,24 @@ bool XSLImportRule::isLoading()
 
-    if (rootSheet)
-        cachedResourceLoader = rootSheet->cachedResourceLoader();
+    RefPtr cachedResourceLoader = rootSheet->cachedResourceLoader();
```
The code just above this checks `rootSheet` for null, but here we dereference the pointer unconditionally. The old code didn’t do that.

rniwa replied:
Yes, this is true but line 99/106 below also unconditionally dereferences the resource loader.

Whilst it's after an early exit for a cyclic relationship, I don't think that's relevant as far as nullify of the resource loader is concerned.

darinadler replied:
We could remove the unnecessary null check above then?

rniwa replied:
I suppose we could. I'll go do that.

rniwa committed:
XSLImportRule and XSLStyleSheet should use weak pointers instead of raw pointers
https://bugs.webkit.org/show_bug.cgi?id=244181

Reviewed by Darin Adler.

Deploy WeakPtr where raw pointers are used.

* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::setXSLStyleSheet):
* Source/WebCore/xml/XSLImportRule.cpp:
(WebCore::XSLImportRule::XSLImportRule):
(WebCore::XSLImportRule::setXSLStyleSheet):
(WebCore::XSLImportRule::loadSheet):
* Source/WebCore/xml/XSLImportRule.h:
(WebCore::XSLImportRule::parentStyleSheet const):
* Source/WebCore/xml/XSLStyleSheet.h:
* Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
(WebCore::XSLStyleSheet::XSLStyleSheet):
(WebCore::XSLStyleSheet::checkLoaded):
(WebCore::XSLStyleSheet::loadChildSheet):
(WebCore::XSLStyleSheet::ownerDocument):
(WebCore::XSLStyleSheet::locateStylesheetSubResource):

Canonical link: https://commits.webkit.org/253695@main
